muxers/mplex/src/lib.rs

```diff
@@ -583,10 +588,23 @@ enum OutboundSubstreamState {
 }
 
 /// Active substream to the remote.
+#[derive(Debug)]
 pub struct Substream {
     /// Substream number.
     num: u32,
     // Read buffer. Contains data read from `inner` but not yet dispatched by a call to `read()`.
     current_data: Bytes,
     endpoint: Endpoint,
 }
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use libp2p_test_muxer::test_muxer;
+
+    test_muxer!(empty_payload, MplexConfig::default());
```

Comment:
I see your point for sure. My thinking was that it'd evolve to do more, to start with it'd take a list of tests, like 

```rust
test_muxer!([empty_payload, one_megabyte, ten_megabytes, bidirectional, interrupted_server, rate_limited_server, interrupted_client, rate_limited_client], Config::default());
test_muxer!([empty_payload, one_megabyte, ten_megabytes, bidirectional], special_config);
``` 

Another motive was that the original issue asked for a single method, `test_muxer()` to be called from implementations, as a kind of black-box almost. That doesn't work so well in the presence of different configuration settings for different muxers and so this was a tentative compromise.

That said, I totally agree that as it stands it hardly carries its weight.

Reply:
I don't think we need multiple configuration settings. All the muxers should be tested the same way. It's the `test_muxer()` itself that would dispatch to the different tests.

Reply:
> All the muxers should be tested the same way. 

This is problematic I think. For some configuration options we could make an effort to normalize settings between impls: mplex has `max_substreams` vs yamux `max_num_streams`, mplex has `max_buffer_size` vs yamux `max_buffer_len`. 
Other things are deeply linked to the specific impl though. Take for instance the difference in behaviour between [`OnReceive`](https://github.com/paritytech/yamux/blob/master/src/lib.rs#L42) and [`OnRead`](https://github.com/paritytech/yamux/blob/master/src/lib.rs#L56).

For these reasons I think it makes sense to be able to run tests using different configs.

> It's the test_muxer() itself that would dispatch to the different tests.

I started out like this and it became annoying to work on the tests dispatched this way; for instance to filter out some tests I have to comment out code. I also don't like not seeing the tests printed out one by one as they run, but maybe that's just me.

Why is it important to have a single function call?

Reply:
I guess it makes sense to have different configurations when you're testing corner cases (which you definitely want to test).
However it would be great if it was as generic as possible, and not rely too much on details that are *too* specific.